aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java

```diff
@@ -222,6 +223,7 @@ protected UploadPartRequest newUploadPartRequest(PartCreationEvent event,
             .withLastPart(event.isLastPart())
             .withUploadId(uploadId)
             .withObjectMetadata(req.getUploadPartMetadata())
+            .withRequestCredentialsProvider(req.getRequestCredentialsProvider())
```

Comment:
Don't we need to add this in `s3#completeMultipartUpload` as well?

Reply:
sorry, can you be more specific on which class this `completeMultipartUpload ` function is being called ?

Reply:
Yeah, sorry for being unclear. For Multipart upload request, we need to do three steps:

- Initiate multipartUpload request
- Upload each part
- Complete multipart upload

I noticed you added the requestCredentialProvider for the first two steps and I was wondering if we need to add it to the third step.

https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L114

Reply:
So If I understand correctly `Complete multipart upload` would be called in the following lines in order

1. [observer.onCompletion(partETags);](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3EncryptionClient.java#L836) 
2. then, [UploadObjectObserver.java#onCompletion](https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L179-L183)
3. [AmazonS3EncryptionClient.java#completeMultipartUpload](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3EncryptionClient.java#L597-L600)
4. [S3CryptoModuleBase.java#completeMultipartUploadSecurely()](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/crypto/S3CryptoModuleBase.java#L377-L401)

above steps 3), 4) and `CompleteMultipartUploadRequest` does not have field to use/set `RequestCredentialsProvider` to their request, so and as far as our use cases that starts with `s3.uploadObject(req)` is using with `MultiFileOutputStream` within the request. 

So, I don't see the needs to change/add `RequestCredentialsProvider` to `Complete multipart upload`.

Reply:
Yeah, `UploadObjectObserver.java#onCompletion` is what I was talking about. (I pasted a wrong link in my previous comment)

It creates a new `CompleteMultipartUploadRequest` request, so it seems odd to me that we only add `RequestCredentialsProvider` for other two requests and not `CompleteMultipartUploadRequest`. 

https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L179-L183

Anyway, thanks for the information. We'll discuss internally of this change and provide an update shortly.

Reply:
We discussed internally and we think we should add `RequestCredentialsProvider` to  `UploadObjectObserver#onCompletion` as well.

Could you confirm adding the requestCredentials to `UploadObjectObserver#onCompletion` would still work in your case?

Reply:
I see your point now, I will try to add `withRequestCredentialsProvider` in `CompleteMultipartUploadRequest` and its caller e.g. [`CompleteMultipartCopy.java#call()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartCopy.java#L73-L103), [`UploadCallable.java#uploadPartsInSeries()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/UploadCallable.java#L242-L277) and [`CompleteMultipartUpload.java#call()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartUpload.java#L84-L112).

I need a bit time to test each use case as well, as I mentioned our code only use `s3.uploadObject(req)` which won't cover all the corner cases. 

meanwhile, it's a bit out of my original scope to fix one problem, it's expended 4 use cases now, I'm wondered if we should on focus on my use case of `s3.uploadObject(req)` here...let me know if you or your team has any thoughts.

Reply:
Yeah, adding it to `UploadObjectObserver#onCompletion` like the code snippet you pasted above should be enough. Thanks!